### ISSUE TITLE
Client: Mitigate BoundProtocol Insanities

### ISSUE THREAD

**Issue by holgerd77** [ethereumjs/ethereumjs-monorepo#3157]
It has become tradition in our team to rant about the BoundProtocol mechanism connecting networking protocols like ETH or SNAP to a client peer. For a very good reason: to call an ETH protocol `GET_BLOCK_HEADERS` message e.g. goes through many (many) hops and it is close to not possible to (easily) follow the code path in a static way (static in a way of: having concrete TypeScript types available which will guide you to the message call implementation).

The original intent of all of this (written by Vinay who did the first client version in JavaScript) was to remain flexible on adding new protocols and therefore bind protocols and runtime and also do the code in a structurally open way (all protocols possible). Especially the latter was more fitting for a JavaScript world but it now hindering in a TypeScript world, since it prevents the most basic type associations (that an RLPX peer (naturally) has an eventual ETH binding e.g.).

I do think having a flexible architecture here has benefits and we should not completely give up on this, neither on the "application" protocol layer (it was e.g. easily possible to add SNAP alongside ETH with this) nor on the transport layer (so: devp2p, since we might *also* want to expand here, and (who knows?) add Portal Network as a network provider here.

When looking at the code I nevertheless think we can nevertheless take various steps to reduce the call chain and make the code readable/followable again. So we can take it there in a first round and see where we go from there. Note that the number of "hops" (getting from the inititial call trigger e.g. for a `GET_BLOCK_HEADERS` message to the actual execution) is really insane (6?? steps or so?) and this is therfore only to *reduce* the number of steps in a first round (to let's say: 2-3) and will not get to 1 in a first round. It should be nevertheless very much possible to do the steps in isolation and get to a point where the (client networking) code becomes readable again.

Following suggested steps:

### Binding

1. The `BoundProtocol` (`net/protocol/boundprotocol.ts`) is handling too much logic to directly replace or refactor (if necessary, also no full picture). We can nevertheless improve typing here by adding very small subclasses "per protocol" (best in the file itself) like the following with several better typed parameters:

```typescript
export class BoundETHProtocol {
  public name: 'ETH' // public name: string
  private protocol: EthProtocol // private protocol: Protocol
  private peer: RlpxPeer // private peer: Peer
  private sender: RlpxSender // private sender: Sender
}
```

This adds little boilerplate which I think is very much worth it.

2. (the more important one)
We completely kill the `public bound: Map<string, BoundProtocol>` property in the generic `Peer` (`src/net/peer.ts`) class. This holds a map of the bound protocol implementations (like `EthProtocol`) where things are ultimately called, as some in-between context. We keep the generic `Peer` class though, and if it's only to reduce complexity on this refactor (and again: *might* also just be worth to keep).

Side note: the suggested changes might draw in some other local refactoring needs which should be manageable though.

(Another side note: even for a meant-to-be flexible implementation this `Map` structure is generally *too* generic)

We then add static optional protocol properties to `RlpxPeer` (`src/net/peer/rlpxpeer.ts`) like the following:

```typescript
class RlpxPeer {
  public ETH?: EthProtocol
  public SNAP?: SnapProtocol
  public LES?: LesProtocol // keep this to reduce complexity
}
```

In `RlpxPeer.bindProtocols()` (so: the "plural" version) we do not call into the upper-class `Peer.bindProtocol()` any more but assign directly to `RlpxPeer.ETH` e.g..

This should make Protocol binding already a lot more straight forward and better typed.

### Calling

So. How is calling done? ...Give me 5 minutes. 😂

So. Took me more like 15 minutes (already through this countless times, haha).

Send stack is the following:
Most (not all) "send triggers" are done in `FullEthereumService` (`src/service/fullethereumservice.ts`) (this is generally not so relevant though).

Here is an example:
1. `peer.eth!.send('BlockHeaders', { reqId, headers: [] })`

The `send()` method is called on the generic `BoundProtocol` right now (and that's why things get lost here), see:

<img width="681" alt="grafik" src="https://github.com/ethereumjs/ethereumjs-monorepo/assets/931137/30185ab2-51ef-4549-8c46-b1897d1a5ceb">

We have this artificial typing (in `peer.ts`) `public eth: (BoundProtocol & EthProtocolMethods) | undefined` which was likely introduced to help here a bit but this can also not completely help here to get over the sadness.

Since we now have this Per-Protocol Bound classes (like `BoundETHProtocol` (or name otherway around, TBD), see above) we likely then also want to introduce simple direct one line wrapper methods in these classes like:

```typescript
sendBlockHeaders(reqId: BigInt, headers: BlockHeader[]) {
  super.send('BlockHeaders', { reqId, headers: [] })
}
```

(Code and typing to be double-checked)

This would then allow the `peer.eth!.send('BlockHeaders', { reqId, headers: [] })` code to be changed in:

```typescript
peer.ETH!.sendBlockHeaders(reqId, headers) // (so: the "new" capital ETH property, see above)
```

This would give us the ability to:

1. Directly call into the "correct" (Bound) protocol class here (so: maybe still a bit "indirect", but the correct protocol association is there) and... (maybe as or more important)
2. Have respectively add proper Per-Message-Types, here for `reqId` and `headers`(at the moment this is all `any`)

So this would bring us already quite a step further on getting to some readable networking code in the client. 🙂

**Comment by acolytec3** [comment]
Will read again this morning but maybe doing what you've proposed along with something like what I described in #2939 is the end result we want?

**Comment by holgerd77** [comment]
> Will read again this morning but maybe doing what you've proposed along with something like what I described in #2939 is the end result we want?

Agree. 🙂

**Comment by jochem-brouwer** [comment]
Quick note, in `boundprotocol` we also have this:

![image](https://github.com/ethereumjs/ethereumjs-monorepo/assets/29359032/6a2f34d6-5613-4d20-9152-6744826f5701)

This is the reason why one can `getBlockHeaders`, but if you CTRL-F `getBlockHeaders this method exists nowhere :sweat_smile: I keep forgetting this.

**Comment by jochem-brouwer** [comment]
The suggestions here have been applied in #3164

Small note: in point (2) you mention 
```typescript
class RlpxPeer {
  public ETH?: EthProtocol
  public SNAP?: SnapProtocol
  public LES?: LesProtocol // keep this to reduce complexity
}
```

I'm assuming you meant the Bound variants of this? (Which are now newly introduced)

**Comment by acolytec3** [comment]
This is where I ultimately hope to land. :crossed_fingers:  
![image](https://github.com/ethereumjs/ethereumjs-monorepo/assets/17355484/f0d40364-fd2e-449b-b141-2bb475b3c337)

#3164 is a good first step though it's really just papering over the underlying issues. But, if it takes us 6 months to refactor the too complex by half integration of `devp2p` and `client.rlpxServer` and then `devp2pRlpxPeer` and `rlpxPeer`, it'll be enough I think.

**Comment by jochem-brouwer** [comment]
I know we discussed this somewhat during our calls, but in your example above how would a peer be picked? How would the logic work? For some peers we do not want to send the new pooled txs (peer already knows these) while for others we want to do this. Is this encapsulated in the logic of that method?

**Comment by acolytec3** [comment]
I'll have to find where it's located  but we have a map of peerIds and the txs already known to them.  I'm pretty sure it's already maintained on a per message basis.

**Comment by holgerd77** [comment]
Closed by #3164
